cryptoadvance/specter-desktop
# Feature: Implement automatic wallet import via Specter URI for MacOS

This PR makes changes to the Electron code of Specter to register the `specter`-protocol with the Specter app and implements one first application of this: automatic wallet import by clicking a URL which contains wallet data as a query parameter. This is currently implemented for MacOS only. The import of the wallet works if Specter is already open or if it is closed.

### Visuals


https://github.com/cryptoadvance/specter-desktop/assets/70536101/e216f53a-5288-4b3c-88c6-9d86659125c2

https://github.com/cryptoadvance/specter-desktop/assets/70536101/c6797298-f584-4f9c-850a-74eca0add76f



### How to test this PR?
```
cd pyinstaller/electron
npx electron-builder build
```
You need to build the app to make the registering of the URI handler work (see https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app). 

Open the app from the `dist` folder. For ARM64 based Macs: `specter-desktop/pyinstaller/electron/dist/mac-arm64/Specter.app`

Open the app to register the URI handler. You also need to set a remote specterd server and run it locally. Best is to use a different port than `25441` otherwise the app is complaining that the address is already in use (not completely sure why, but this is for dev only anyways).

So if you run for example:
`python -m cryptoadvance.specter server --debug --config DevelopmentConfig --port 25442`

You have to set in the Electron app:
<img width="808" alt="image" src="https://github.com/cryptoadvance/specter-desktop/assets/70536101/dcb56678-3d62-478a-82b9-717f7d8e1aea">

The last step closes the app. 

Now you can either directly use this test URL to import a multisig wallet or open the app and then click on the link in a browser tab:

[specter://?action=importWallet&data=%7B%22label%22%3A%22Swan%20Vault%20(created%20via%20Specter)%22%2C%22blockheight%22%3A0%2C%22descriptor%22%3A%22wsh(sortedmulti(2%2C%5B7737acb8%5DtpubD6NzVbkrYhZ4X3vCdevm3aFUdeSAWsbR4UhhdApbXnVUnvuwtq5NYqQ7zCkNhYG4JcPdGW4Wg4ZrVSYGpdoWsn337qmkP6CFgs7pZid12Sa%2F0%2F*%2C%5B8c24a510%2F48h%2F1h%2F0h%2F2h%5DtpubDDzWqfZ5TH48383Byd9PFGxEP1Ws5NVXyYcHTmnHwmhJciowLeBDWNHcpLGocofanSyVHeiNqL4HZkXZfKM7NKm7gZZoPjmA9vTKPpwRSkx%2F0%2F*%2C%5Bdcbf0caf%2F48h%2F1h%2F0h%2F2h%5DtpubDFd7VxopNeZg93uqR7CSvJLkw3UanF8rywdQTxhCPFt1P33eZkxJJ91XXEbY2Q4Suw3jyscRwGzjVyfgq97N7sRvPHQVxruHwsKvsvQizSk%2F0%2F*))%23g890ye6p%22%2C%22devices%22%3A%5B%7B%22type%22%3A%22other%22%2C%22label%22%3A%22Cloud%20Key%20Zoo%20Where%22%7D%2C%7B%22type%22%3A%22jade%22%2C%22label%22%3A%22Jade%20Ghost%20Machine%22%7D%2C%7B%22type%22%3A%22jade%22%2C%22label%22%3A%22Jade%20Keen%20Join%22%7D%5D%7D](specter://?action=importWallet&data=%7B%22label%22%3A%22Swan%20Vault%20(created%20via%20Specter)%22%2C%22blockheight%22%3A0%2C%22descriptor%22%3A%22wsh(sortedmulti(2%2C%5B7737acb8%5DtpubD6NzVbkrYhZ4X3vCdevm3aFUdeSAWsbR4UhhdApbXnVUnvuwtq5NYqQ7zCkNhYG4JcPdGW4Wg4ZrVSYGpdoWsn337qmkP6CFgs7pZid12Sa%2F0%2F*%2C%5B8c24a510%2F48h%2F1h%2F0h%2F2h%5DtpubDDzWqfZ5TH48383Byd9PFGxEP1Ws5NVXyYcHTmnHwmhJciowLeBDWNHcpLGocofanSyVHeiNqL4HZkXZfKM7NKm7gZZoPjmA9vTKPpwRSkx%2F0%2F*%2C%5Bdcbf0caf%2F48h%2F1h%2F0h%2F2h%5DtpubDFd7VxopNeZg93uqR7CSvJLkw3UanF8rywdQTxhCPFt1P33eZkxJJ91XXEbY2Q4Suw3jyscRwGzjVyfgq97N7sRvPHQVxruHwsKvsvQizSk%2F0%2F*))%23g890ye6p%22%2C%22devices%22%3A%5B%7B%22type%22%3A%22other%22%2C%22label%22%3A%22Cloud%20Key%20Zoo%20Where%22%7D%2C%7B%22type%22%3A%22jade%22%2C%22label%22%3A%22Jade%20Ghost%20Machine%22%7D%2C%7B%22type%22%3A%22jade%22%2C%22label%22%3A%22Jade%20Keen%20Join%22%7D%5D%7D )

The structure of the wallet data is (in JSON format, the above is URI encoded) - which is taken from the current export standard of Specter when you use this (the blockheight is optional):

<img width="690" alt="image" src="https://github.com/cryptoadvance/specter-desktop/assets/70536101/d9012dbd-7a1a-4e4b-8f97-23e82cc96d8c">


```
{
    "label": "Super Duper Wallet",
    "blockheight": 0,
    "descriptor": someDescriptor,
    "devices": [
        {
            "type": "trezor",
            "label": "Key 1"
        },
        {
            "type": "jade",
            "label": "Key 2"
        },
        {
            "type": "coldcard",
            "label": "Key 3"
        }
    ]
}
```
